content/shmem_alltoall_nb.tex

```diff
@@ -0,0 +1,129 @@
+  respectively.
+}
+\apiargument{IN}{tag}{A user defined tag to order the collective operation;
+SHMEM\_COLL\_UNORDERED can be provided if no order is required.}
+\apiargument{OUT}{request}{An opaque request handle identifying the collective
+operation.}
+
+\end{apiarguments}
+
+\apidescription{
+    The \FUNC{shmem\_alltoall\_nb} routines are collective routines. All
+    \acp{PE} in the provided team must participate in the collective. If
+    \VAR{team} compares equal to \LibConstRef{SHMEM\_TEAM\_INVALID} or is
+    otherwise invalid, the behavior is undefined.
+
+    {\bf Invocation and completion}: A call to the nonblocking alltoall routine posts the operation and returns
+    immediately without necessarily completing the operation. On the successful
+    post of the operation, an opaque request handle is created and returned. The
+    operation is completed after a call to \FUNC{shmem\_req\_test} or
+    \FUNC{shmem\_req\_wait}. When the operation is complete, the request handle
+    is deallocated and cannot be reused.
+
+    Though nonblocking alltoall varies in invocation and completion semantics
+    when compared to blocking alltoall, the data exchange semantics are similar.
+
+    {\bf Data exchange semantics}:
+    In this routine, each \ac{PE}
+    participating in the operation exchanges \VAR{nelems} data elements
+    with all other \acp{PE} participating in the operation.
+    The size of a data element is:
+    \begin{itemize}
+    \item 32 bits for \FUNC{shmem\_alltoall32}
+    \item 64 bits for \FUNC{shmem\_alltoall64}
+    \item 8 bits for \FUNC{shmem\_alltoallmem}
+    \item \FUNC{sizeof}(\TYPE{}) for alltoall routines taking typed \VAR{source} and \VAR{dest}
+    \end{itemize}
+
+    The data being sent and received are
+    stored in a contiguous symmetric data object. The total size of each \ac{PE}'s
+    \VAR{source} object and \VAR{dest} object is \VAR{nelems} times the size of
+    an element
+    times \VAR{N}, where \VAR{N} equals the number of \acp{PE} participating
+    in the operation.
+    The \VAR{source} object contains \VAR{N} blocks of data
+    (where the size of each block is defined by \VAR{nelems}) and each block of data
+    is sent to a different \ac{PE}.
+
+    The same \dest{} and \source{}
+    arrays, and same value for nelems
+    must be passed by all \acp{PE} that participate in the collective.
+
+    Given a \ac{PE} \VAR{i} that is the \kth \ac{PE}
+    participating in the operation and a \ac{PE}
+    \VAR{j} that is the \lth \ac{PE}
+    participating in the operation,
+
+    \ac{PE} \VAR{i} sends the \lth block of its \VAR{source} object to
+    the \kth block of
+    the \VAR{dest} object of \ac{PE} \VAR{j}.
+
+    
+    Like data exchange semantics, the entry and completion
+    criteria of blocking and nonblocking alltoall is similar. 
+
+    {\bf Entry criteria}: Before any \ac{PE} calls a \FUNC{shmem\_alltoall\_nb} routine,
```

Comment:
This entry criteria is confusing - if we have the same entry criteria as blocking operations - then the users are supposed to make sure the src/dst buffers are available at the point of entry - meaning the users are expected to sync with themselves before launching the nb collective operation.

AFAIU - the src/dst needs to be available when everyone in the team reaches the state - which can be determined during runtime - without the need for explicit sync before launching the nb collective operation.

Please clarify the intended semantics.

Reply:
Naveen, my intent is to keep it consistent with the semantics of blocking collectives. I wanted to improve the description. I don’t see how the current wording is violating that. If the wording doesn’t reflect that, then we can fix it.